### PR TITLE
Contributor's Guide draft

### PR DESCRIPTION
**Pull request by lilianabs**
@logankilpatrick, @DhairyaLGandhi can you please have a look at this first draft? Some sections are incomplete and it'd be very helpful if you could give me some feedback on the current content :)

### CODE REVIEW

**Comment by logankilpatrick** [comment]
This is an awesome first pass!! Nothing big jumps out at me as having to be addressed so far. Any additional thoughts @DhairyaLGandhi ?

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -1,23 +1,155 @@
...
-## Write Tutorials
+Contributing to Flux brings several benefits to you and the community such as:
 
-Tutorials are an important part of making Deep Learning and Flux accessible to everyone. We already have some great ones on our website here: https://fluxml.ai/tutorials.html and welcome new ones, especially if they are beginner focused, of which beginners tend to write the best ones. A good first step is to find a tutorial you find helpful from another ecosystem and try to rewrite it using Flux (making sure to properly attribute and not plagiarize)
+* Contribute to the cause of open and reproducible science.
+* Help grow the adoption of the FluxML ecosystem.
+* Become a member of a community that’s excited about open source and sharing knowledge.
+* Polish your teaching skills by helping others get started using Flux.
+* Showcase your work by creating examples that others can use in their projects.
+* Build a track record of public contributions which will help build your career.
```
Comment (darsnack) [inline CONTRIBUTING.md:24]:
```suggestion
* Contribute to the cause of open and reproducible science
* Help grow the adoption of the FluxML ecosystem
* Become a member of a community that’s excited about open source and sharing knowledge
* Polish your teaching skills by helping others get started using Flux
* Showcase your work by creating examples that others can use in their projects
* Build a track record of public contributions which will help build your career
```

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -1,23 +1,155 @@
...
+
+## Learn Flux
+
+If you need to learn about Julia and Flux, then you can check out the following resources:
+
+* [JuliaAcademy](https://juliaacademy.com/) introductory courses to Julia and Flux.
+* Flux’s official [Getting Started](https://fluxml.ai/getting_started.html) tutorial.
+* [Deep Learning with Flux - A 60 Minute Blitz](https://fluxml.ai/tutorials/2020/09/15/deep-learning-flux.html). This is a quick intro to Flux loosely based on PyTorch’s tutorial.
+* [Flux Model Zoo](https://github.com/FluxML/model-zoo) showcases various demonstrations of models that you can reuse with your own data.
+* [Flux’s official documentation](https://fluxml.ai/Flux.jl/stable/).
```
Comment (darsnack) [inline CONTRIBUTING.md:48]:
```suggestion
* [JuliaAcademy](https://juliaacademy.com/) introductory courses to Julia and Flux
* Flux’s official [Getting Started](https://fluxml.ai/getting_started.html) tutorial
* [Deep Learning with Flux - A 60 Minute Blitz](https://fluxml.ai/tutorials/2020/09/15/deep-learning-flux.html): a quick intro to Flux loosely based on PyTorch’s tutorial
* [Flux Model Zoo](https://github.com/FluxML/model-zoo) showcases various demonstrations of models that you can reuse with your own data
* [Flux’s official documentation](https://fluxml.ai/Flux.jl/stable/)
```

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -1,23 +1,155 @@
...
+## Get help
+
+The Flux community is more than happy to help you with any questions that you might have related to your contribution. You can get in touch with the community in any of the following channels:
+
+* **Stack Overflow:** Search for [Flux/ ML](https://stackoverflow.com/questions/tagged/flux.jl) tags.
+* **Discourse forum:** Go to [Machine Learning in Julia community](https://discourse.julialang.org/c/domain/ml/24).
+* **Events:** Attend the [ML bi-weekly calls](https://julialang.org/community/#events).
+* **Community team:** Join the group of [community maintainers](https://github.com/FluxML/ML-Coordination-Tracker) supporting the FluxML ecosystem.
+* **Slack:** Join the [Official Julia Slack](https://julialang.org/slack/) for casual conversation.
+* **Zulip:** Join the [Zulip Server for the Julia programming language](https://julialang.zulipchat.com/login/) community.
```
Comment (darsnack) [inline CONTRIBUTING.md:59]: Just reordering based on importance and added link to Zulip stream

```suggestion
* **Discourse forum:** Go to [Machine Learning in Julia community](https://discourse.julialang.org/c/domain/ml/24)
* **Community team:** Join the group of [community maintainers](https://github.com/FluxML/ML-Coordination-Tracker) supporting the FluxML ecosystem (see the [Zulip stream](https://julialang.zulipchat.com/#narrow/stream/237432-ml-ecosystem-coordination) as well)
* **Slack:** Join the [Official Julia Slack](https://julialang.org/slack/) for casual conversation (see `#flux-bridged` and `#machine-learning`)
* **Zulip:** Join the [Zulip Server for the Julia programming language](https://julialang.zulipchat.com/login/) community
* **Stack Overflow:** Search for [Flux/ ML](https://stackoverflow.com/questions/tagged/flux.jl) tags
* **Events:** Attend the [ML bi-weekly calls](https://julialang.org/community/#events)
```

**Review by darsnack — COMMENTED** [review]
Just some minor changes. This is amazing, thank you!

**Comment by DhairyaLGandhi** [comment]
I get the sense that we need to give more emphasis to writing tutorials and models in addition to the core functionality. We should also talk about the information to be included in an issue and PR and how to present it best. One is to limit PRs to make only the necessary changes, and write motivating tests before hand.

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -1,23 +1,155 @@
...
 
-One of the best ways to contribute is by looking at issues labeled ["help wanted"](https://github.com/FluxML/Flux.jl/labels/help%20wanted). These issues are not always very beginner friendly but you are welcome to ask clarifying questions or just browse them to see if there is anything that seems interesting to help with. 
+Flux has always had the mission of being a simple, hackable and performant approach to machine learning, which is extended to a number of domains in science by means of differentiable programming. 
 
-## Good First Issues
+Julia provides features that Machine Learning researchers and practitioners need. However, the ecosystem is not yet mature enough so that developers can rewrite it in Julia. This provides a great opportunity for you to contribute to the [future of Machine Learning and the Julia language in general](https://towardsdatascience.com/the-future-of-machine-learning-and-why-it-looks-a-lot-like-julia-a0e26b51f6a6).
```
[the remark this answers is not in the document]
Comment (DhairyaLGandhi) [reply CONTRIBUTING.md:9]: I would reword it to say the ecosystem is rapidly growing, but maybe we can remove the links to the blog posts. Is it common for contribution guides to have such links in them?

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -1,23 +1,155 @@
...
 
-## Other ideas
+## Tools
 
-We are open to contributions in all shapes and forms! If you have an idea, please suggest it and we will do our best to help you bring it to life!
+Before you start contributing to Flux, you need the following tools:
+
+* **Julia:** For more information on how to install Julia, see [Download Julia (in under 2.5 minutes)](https://www.youtube.com/watch?v=t67TGcf4SmM).
+* **IDE:** You can set one of the following IDEs as well as their extensions for developing in Julia:
+   * [Julia for Visual Studio Code](https://www.julia-vscode.org/)
```
[the remark this answers is not in the document]
Comment (DhairyaLGandhi) [reply CONTRIBUTING.md:32]: I don't think we need to say this here.

File: CONTRIBUTING.md (hunk 3 above)
Comment (DhairyaLGandhi) [reply CONTRIBUTING.md:59]: Pretty sure the "Get Help" section should start with places to get existing help from. In these cases Discourse and SO would be more important followed by Slack, Zulip and then the ML Tracker.

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -1,23 +1,155 @@
...
+
+Flux.jl and its ecosystem follow the ColPrac: Contributor's Guide on Collaborative Practises for Community Packages guide. For more information about best practises on contributing to packages, see the [ColPrac guide](https://github.com/SciML/ColPrac).
+
+We also suggest taking a look at the [Julia Ecosystem Contributor’s Guide](https://julialang.org/contribute/) guide which goes over higher-level topics like why you should contribute to the Julia ecosystem, different contribution pathways, and more!
+
+In this section, you can find information on the following:
+
+* [How to file a bug report]()
+* [How to submit a contribution]()
+* [Flux source code organisation]()
```
Comment (DhairyaLGandhi) [inline CONTRIBUTING.md:124]: Should link to the appropriate sections.

**Review by DhairyaLGandhi — COMMENTED** [review]
Thanks for tackling this! This is shaping up really well. I've left a few thoughts on the document so far.

File: CONTRIBUTING.md (hunk 3 above)
Comment (darsnack) [reply CONTRIBUTING.md:59]: The reason I put SO lower is cause no devs seem to consistently check it. So I ordered based on how quickly they might get help from fellow contributors.

File: CONTRIBUTING.md (hunk 6 above)
Comment (lilianabs) [reply CONTRIBUTING.md:124]: Links fixed. Just a question: does it makes sense to have the section Flux source code organisation? I haven't finished that section and just wondering if it's worth it.

**Comment by lilianabs** [comment]
@logankilpatrick, @DhairyaLGandhi, @darsnack I've added the suggestions. Could you please have a look at the section Flux source code organization and comment on it. Do you think is worth having it?

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -1,23 +1,158 @@
...
+
+### Flux source code organization
+
+The following table shows how the Flux code is organized:
+
+| Directory  | Contents |
+| ------------- | ------------- |
+| docs  | Documentation site  |
+| paper  | Paper that describes Flux |
+| perf  |    |
```
Comment (darsnack) [inline CONTRIBUTING.md:156]: For now I would leave this out, since it is kind of a dead branch.

File: CONTRIBUTING.md [hunk 8]
```diff
@@ -1,23 +1,158 @@
...
+
+Before you begin working on your contribution, make sure you have installed Julia and an IDE (see section [Tools](#tools)). In case of any issues, remember you can always [get help](#get-help) from the community. 
+
+To make your first contribution to Flux:
+
+1. Read about getting started with **Git** and **how to contribute to an open-source project**. The tutorial [Making a first Julia pull request](https://kshyatt.github.io/post/firstjuliapr/) is a great starting point for learning about git and how to contribute to the Julia language in general. 
+2. Read the [Contribution ideas](#contribution-ideas) section to get inspiration.
+3. Set up your local environment. Depending on your contribution, fork and clone the Flux, Model Zoo or Flux website GitHub repositories. 
+4. Build Flux locally.
+5. Familiarise yourself with the Flux source code. For more information on the source code organisation, see section….
```
Comment (darsnack) [inline CONTRIBUTING.md:72]:
```suggestion
5. Familiarise yourself with the Flux source code. See [Flux source code organization](#flux-source-code-organization) for more information.
```

**Review by darsnack — CHANGES_REQUESTED** [review]
Looks good to me!

File: CONTRIBUTING.md (hunk 7 above)
Comment (lilianabs) [reply CONTRIBUTING.md]: Thank you so much @darsnack for all of your comments!

Comment (darsnack) [reply CONTRIBUTING.md]: No problem! Just to clarify, I meant delete the row for `perf` only. I think the section you added on source code organization is a great idea.

Comment (lilianabs) [reply CONTRIBUTING.md]: Ah! Sorry for the confusion. I added the table :)

**Review by lilianabs — COMMENTED** [review]
Updated :)

**Review by darsnack — APPROVED** [review]
Thank you @lilianabs!
